oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,293 @@
+#################################
+OEP-0019: Developer Documentation
+#################################
+
+.. list-table::
+  
+   * - OEP
+     - :doc:`OEP-0019 </oeps/oep-0019>`
+   * - Title
+     - Developer Documentation
+   * - Last Modified
+     - 2018-11-28
+   * - Authors
+     - Grant Goodman, Robert Raposa, Nimisha Asthagiri
+   * - Arbiter
+     - Cali Stenson
+   * - Status
+     - Draft
+   * - Type
+     - Best Practice
+   * - Created
+     - 2018-03-27
+   * - Review Period
+     - 2018-12-06 - 2018-12-20
+
+.. image:: oep-0019/write.png
+   :align: right
+
+.. contents::
+   :local:
+   :depth: 2
+
+Context
+-------
+
+Developers who work on the Open edX platform, both inside and outside of edX, need accurate and current documentation
+about platform architecture, APIs, and development best practices.
+
+Developer documentation is inconsistent and poorly maintained. Documents are also difficult to find. They
+exist in multiple locations, including `Read the Docs`_, GitHub_ repositories, Confluence_ wiki, and `Google Docs`_.
+
+This proposal focuses on long-term developer *Documentation* needs. It does not concern itself with any transitory
+*Notes* that a development team may use for project management, brainstorming, meeting notes, etc. Code *Comments*
+that a developer may add to clarify or rationalize a code segment are also out of scope of this OEP.
```

Comment:
In my recent experience, we've been updating our first ADR in a newish project during the course of development, instead of creating a new record each time we change something, since things are in a very fluid state at the moment.  As the pace of development slows, or as we move toward a second milestone of the project, I expect that we'll eventually "freeze" the initial ADR.  When changes/decisions are made in the future, they should be far less frequent than they are now, or pertain to a later project milestone.  Those changes can have their own ADRs.

I've found that there is some overlap between things I've written in ADRs and code comments - the comments tend to be a little more technical and focused to the surrounding code, without having the big picture context.  IMO, the surrounding context is captured by the ADR.

ADR: "What is the decision and why did we make it?"
Comment: "Explain the nuances of this small set of code and briefly how it relates to your decision."

Reply:
@iloveagent57 Good comparison between ADRs and code comments.  I added that as a note in the OEP.
@johnbaldwin I added a link to Nygard's blog-post on ADRs and wrote out the acronym.  Thanks.

Reply:
@nasthagiri Is it possible for us to develop our own definition of ADRs and when they are appropriate rather than relying on an external resource?

Also, I am still looking, but do we have an example template ADR and complete ADR?

Reply:
Just read further on, looks like that was also added.